bad-bits/lib/store.js

```diff
@@ -1,58 +1,106 @@
  */
 export async function updateBadBitsDatabase(env, currentHashes, etag) {
   const startedAt = Date.now()
-  const timestamp = new Date().toISOString()
-  const insertBadBitStmt = env.DB.prepare(
-    `
-    INSERT INTO bad_bits (hash, last_modified_at) VALUES (?, ?)
-    ON CONFLICT(hash) DO UPDATE SET last_modified_at = excluded.last_modified_at
-    `,
-  )
 
-  let updated = 0
-  const remainingHashes = Array.from(currentHashes)
-  while (remainingHashes.length > 0) {
-    // pop first N hashes from remainingHashes
-    const batchHashes = remainingHashes.splice(0, 10_000)
+  console.log('getting latest hashes')
+  const oldHashes = new Set(await getAllBadBitHashes(env))
+
+  const addedHashes = []
+  const removedHashes = []
 
-    await env.DB.batch(
-      batchHashes.map((hash) => insertBadBitStmt.bind(hash, timestamp)),
+  console.log('iterating current hashes')
+  for (const hash of currentHashes) {
+    if (!oldHashes.has(hash)) {
+      addedHashes.push(hash)
+    }
+    if (addedHashes.length >= MAX_TOTAL_CHANGES) {
+      break
+    }
+  }
+  for (const hash of oldHashes) {
+    if (!currentHashes.has(hash)) {
+      removedHashes.push(hash)
+    }
+    if (addedHashes.length + removedHashes.length >= MAX_TOTAL_CHANGES) {
+      break
+    }
+  }
+
+  console.log('writing added hashes')
+  for (let i = 0; i < Math.ceil(addedHashes.length / MAX_KV_BATCH_SIZE); i++) {
+    const batch = addedHashes.slice(
+      i * MAX_KV_BATCH_SIZE,
+      (i + 1) * MAX_KV_BATCH_SIZE - 1,
+    )
+    await Promise.all(
+      batch.map((hash) => env.KV.put(`bad-bits:${hash}`, 'true')),
     )
+  }
 
-    updated += batchHashes.length
-    console.log(
-      `Inserted/updated ${updated} bad bits in ${Date.now() - startedAt}ms`,
+  console.log('deleting removed hashes')
+  for (
+    let i = 0;
+    i < Math.ceil(removedHashes.length / MAX_KV_BATCH_SIZE);
+    i++
+  ) {
+    const batch = removedHashes.slice(
+      i * MAX_KV_BATCH_SIZE,
+      (i + 1) * MAX_KV_BATCH_SIZE - 1,
     )
+    await Promise.all(batch.map((hash) => env.KV.delete(`bad-bits:${hash}`)))
   }
 
-  const statements = [
-    env.DB.prepare('DELETE FROM bad_bits WHERE last_modified_at < ?').bind(
-      timestamp,
-    ),
-  ]
+  console.log('storing latest hashes')
+  const latestHashes = new Set(oldHashes)
+  for (const hash of addedHashes) {
+    latestHashes.add(hash)
+  }
+  for (const hash of removedHashes) {
+    latestHashes.delete(hash)
+  }
 
-  if (etag) {
-    statements.push(
-      env.DB.prepare(
-        'INSERT INTO bad_bits_history (timestamp, etag) VALUES (?, ?)',
-      ).bind(timestamp, etag),
+  for (
+    let i = 0;
+    i < Math.ceil(latestHashes.size / KV_SEGMENT_MAX_HASH_COUNT);
+    i++
+  ) {
+    await env.KV.put(
+      `bad-bits:_latest-hashes:${i}`,
+      [...latestHashes]
+        .slice(
+          i * KV_SEGMENT_MAX_HASH_COUNT,
+          (i + 1) * KV_SEGMENT_MAX_HASH_COUNT - 1,
```

Comment:
```suggestion
          (i + 1) * KV_SEGMENT_MAX_HASH_COUNT,
```

Same off-by-one error here.